dm/pkg/shardddl/optimism/keeper.go

```diff
@@ -20,79 +20,67 @@ import (
-func NewLockKeeper() *LockKeeper {
+func NewLockKeeper(getDownstreamMetaFunc func(string) (*config.DBConfig, string)) *LockKeeper {
 	return &LockKeeper{
-		locks: make(map[string]*Lock),
+		locks:                 make(map[string]*Lock),
+		downstreamMetaMap:     make(map[string]*DownstreamMeta),
+		getDownstreamMetaFunc: getDownstreamMetaFunc,
 	}
 }
 
-// RebuildLocksAndTables rebuild the locks and tables.
-func (lk *LockKeeper) RebuildLocksAndTables(
-	cli *clientv3.Client,
-	ifm map[string]map[string]map[string]map[string]Info,
-	colm map[string]map[string]map[string]map[string]map[string]DropColumnStage,
-	lockJoined map[string]schemacmp.Table,
-	lockTTS map[string][]TargetTable,
-	missTable map[string]map[string]map[string]map[string]schemacmp.Table,
-) {
-	var (
-		lock *Lock
-		ok   bool
-	)
-	for task, taskInfos := range ifm {
-		for source, sourceInfos := range taskInfos {
-			for schema, schemaInfos := range sourceInfos {
-				for table, info := range schemaInfos {
-					lockID := utils.GenDDLLockID(info.Task, info.DownSchema, info.DownTable)
-					if lock, ok = lk.locks[lockID]; !ok {
-						lock = NewLock(cli, lockID, info.Task, info.DownSchema, info.DownTable, lockJoined[lockID], lockTTS[lockID])
-					}
-					// filter info which doesn't have SourceTable
-					// SourceTable will be changed after user update block-allow-list
-					// But old infos still remain in etcd.
-					// TODO: add a mechanism to remove all outdated infos in etcd.
-					if !lock.TableExist(info.Source, info.UpSchema, info.UpTable) {
-						delete(ifm[task][source][schema], table)
-						continue
-					}
-					lk.locks[lockID] = lock
-					lock.tables[info.Source][info.UpSchema][info.UpTable] = schemacmp.Encode(info.TableInfoBefore)
-					if columns, ok := colm[lockID]; ok {
-						lock.columns = columns
-					}
-				}
-			}
-		}
+// SetDropColumns set drop columns for lock keeper.
+func (lk *LockKeeper) SetDropColumns(dropColumns map[string]map[string]map[string]map[string]map[string]DropColumnStage) {
+	lk.dropColumns = dropColumns
+}
+
+// getDownstreamMeta gets and cached downstream meta.
+func (lk *LockKeeper) getDownstreamMeta(task string) (*DownstreamMeta, error) {
+	if downstreamMeta, ok := lk.downstreamMetaMap[task]; ok {
+		return downstreamMeta, nil
 	}
 
-	// update missTable's table info for locks
-	for lockID, lockTable := range missTable {
-		for source, sourceTable := range lockTable {
-			for schema, schemaTable := range sourceTable {
-				for table, tableinfo := range schemaTable {
-					if _, ok := lk.locks[lockID]; !ok {
-						continue
-					}
-					if !lk.locks[lockID].TableExist(source, schema, table) {
-						continue
-					}
-					lk.locks[lockID].tables[source][schema][table] = tableinfo
-				}
-			}
-		}
+	dbConfig, meta := lk.getDownstreamMetaFunc(task)
+	if dbConfig == nil {
+		return nil, terror.ErrMasterOptimisticDownstreamMetaNotFound.Generate(task)
 	}
+	db, err := conn.DefaultDBProvider.Apply(dbConfig)
```

Comment:
defer db.Close()

Reply:
close when remove

Reply:
how about pass dbConfig into DownstreamMeta and lazy open conection ?

Reply:
ece233f